Framework/Geometry/src/Instrument/ComponentInfo.cpp

```diff
@@ -239,6 +241,167 @@ double ComponentInfo::solidAngle(const size_t componentIndex,
+  const auto innerRangeDet = m_componentInfo->detectorRangeInSubtree(index);
+  auto nSubComponents = innerRangeComp.end() - innerRangeComp.begin() - 1;
+  auto nSubDetectors =
+      std::distance(innerRangeDet.begin(), innerRangeDet.end());
+  auto nY = nSubDetectors / nSubComponents;
+  size_t bottomLeft = *innerRangeDet.begin();
+  size_t topRight = bottomLeft + nSubDetectors - 1;
+  size_t topLeft = bottomLeft + (nY - 1);
+  size_t bottomRight = topRight - (nY - 1);
+
+  mutableBB.grow(componentBoundingBox(bottomLeft, reference));
+  mutableBB.grow(componentBoundingBox(topRight, reference));
+  mutableBB.grow(componentBoundingBox(topLeft, reference));
+  mutableBB.grow(componentBoundingBox(bottomRight, reference));
+
+  // Get bounding box for rectangular bank.
+  // Record detector ranges to skip
+  mutableDetExclusions.emplace(std::make_pair(bottomLeft, topRight));
+  // Skip all sub components.
+  mutableIterator = innerRangeComp.rend();
+}
+
+/**
+ * Grow the bounding box on the basis that the component described by index is a
+ * bank containing only detector tubes. Only tube tips need to be treated.
+ *
+ * For each tube, find the start and end index
+ * Grow the bounding box by the top and bottom of the tube, for every tube.
+ * Record detectors that form part of this bank that are enclosed and therefore
+ * already inclusive of the external bounding box Based on the number of
+ * sub-component tubes, which are also fully enclosed, skip the component
+ * iterator forward so that those will not need to be evaluated.
+ *
+ * @param index : Index of the component to get the bounding box for
+ * @param reference : Reference bounding box (optional)
+ * @param mutableBB : Output bounding box. This will be grown.
+ * @param mutableDetExclusions : Output detector exclusions to append to. These
+ * are ranges of detector indices that we do NOT need to consider for future
+ * bounding box calculations for detectors.
+ * @param mutableIterator : Iterator to the next non-detector component.
+ */
+template <typename IteratorT>
+void ComponentInfo::growBoundingBoxAsBankOfTubes(
+    size_t index, const BoundingBox *reference, BoundingBox &mutableBB,
+    std::stack<std::pair<size_t, size_t>> &mutableDetExclusions,
+    IteratorT &mutableIterator) const {
+  auto innerRangeComp = m_componentInfo->componentRangeInSubtree(index);
+  auto nSubComponents = innerRangeComp.end() - innerRangeComp.begin() - 1;
+  auto innerRangeDet = m_componentInfo->detectorRangeInSubtree(index);
+  auto nSubDetectors =
+      std::distance(innerRangeDet.begin(), innerRangeDet.end());
+  auto nY = nSubDetectors / nSubComponents;
+  size_t bottomIndex = *innerRangeDet.begin();
+  size_t lastIndex = *(innerRangeDet.end() - 1);
+  while (bottomIndex < lastIndex) {
+    auto topIndex = bottomIndex + (nY - 1);
+    mutableBB.grow(componentBoundingBox(bottomIndex, reference));
+    mutableBB.grow(componentBoundingBox(topIndex, reference));
```

Comment:
I'll try to make that refactor.

Reply:
Looking at the implementations I think it's probably best left as it is. Unless it is determined that there really are situations where we have banks of tubes where the number of pixels in the tubes is not  the same across the bank. 

The current mechanism should be slightly faster, with only one call to `ComponentInfo::detectorRangeInSubtree` rather than one call per tube.

Reply:
Isn't `detecRangeInSubtree` basically zero cost?

Reply:
Yes. I'll see if making the change has any effect on overall performance.

Reply:
@SimonHeybrock unfortunately I can't do this without degrading performance to some extent. The reason is that the current implementation allows us to ignore all detectors that form part of that bank in one go. Trying the alternative method requires us to store and evaluate many more individual ranges of detectors to exclude from detector bounding box calculations. I did however just slightly rewrite the `getBoundingBoxByDetectors` method to avoid deallocations of the map.